Fix eof() and read*() behaviour for ::File and ::LibuvStream

samoconnor opened the pull request:
See [new test at end of test/file.jl](https://github.com/samoconnor/julia/commit/8f0ef7f87900b3bf5592a724ad38ecb9703d5f70#diff-ebb9c6f01d71511898915e3e728b9917R1135)

See new [test/read.jl](https://github.com/samoconnor/julia/blob/uv_fs_read_eof/test/read.jl)

samoconnor committed:
Fix uv_fs_read() EOF handling.

samoconnor committed:
New test/read.jl

Add eof(::File), seek(::File) etc

samoconnor commented:
It looks like [this new test](https://github.com/samoconnor/julia/commit/f7da086547a2626587e95c455ae9dcc1271d6f8b#diff-435745fdfa02a0dc4f3b747ba977af11R68) is [failing on Win32](https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.12647/job/e56svtn2n1tweb9a#L1177)

```
	From worker 2:	  Expression: readall(io()) == text
	From worker 2:	   Evaluated: "C1,C2 1,2 a,b" == "C1,C2\n1,2\na,b\n"
```
Maybe a DOS line-ending issue?

Could someone who has windows please look at this?

tkelman commented:
Please separate the introduction of new methods from bug fixes into separate PRs.

samoconnor commented:
Appveyor now says  `rmdir: Directory not empty` after a `mktempdir() do dir ... end` block ??

samoconnor commented:
> separate the introduction of new methods from bug fixes

Adding the missing methods _is_ the bug fix.

Handling `uv_fs_read()` EOF requires `eof(::File)` because calling `eof(::File)` is a precondition of calling `read(::File, ::UInt8)`. That in-turn requires `skip(::File)`, and you can't really have `skip(::File)` without having `seek(::File)` and `seekend(::File)` as well.

tkelman commented:
> Adding the missing methods is the bug fix.

Then what was the purpose of opening the PR with a single commit, then adding additional content to it? The new test in the first commit didn't seem to need these new methods to pass, so why are these in the same PR?

samoconnor committed:
Fix read!(::LibuvStream) and readbytes!(::LibuvStream)
to be consisitent with other ::IOs
See test case

samoconnor commented:
@vtjnash, does this fix look right to you?

vtjnash commented on test/read.jl:
```diff
@@ -0,0 +1,245 @@
+
+# TCPSocket
+
+# PR#14627
```
just merged this

samoconnor replied:
Nice one. I've just rebased.

vtjnash commented on test/read.jl:
```diff
@@ -0,0 +1,245 @@
+@windows ? nothing : begin
+
+# PipeEndpoint
+socketname = joinpath(dir, "socket")
```
this test should work on windows, but the socket pipes need to be in the right namespace (`"\\\\.\\pipe\\<socketname>"`, i think)

samoconnor replied:
I think you're right,
However, I haven't done any serious dev work on Windows for about 15 years, so my memory of it isn't good enough to get it working. I went though a bunch of cycles of push, wait for appveyor, guess, push again... I managed to work around `mktempdir() do ...` being broken on windows, but appveyor is just way too slow for debugging the sockets stuff.

I've [added a note in the test code](https://github.com/samoconnor/julia/commit/0032cd62b397f3917c94380791e0559e73ce8e28#diff-435745fdfa02a0dc4f3b747ba977af11R46) to the effect that this stuff should work on windows but that someone who has the right knowledge and tools will need to look at it.

vtjnash commented on base/filesystem.jl:
```diff
@@ -192,6 +216,18 @@ function position(f::File)
+    ret = ccall(:jl_fs_read, Int32, (Int32, Ptr{Void}, Csize_t), f.handle, a, 1)
+    systemerror("eof", ret == -1)
+    if ret == 1
+        skip(f, -1)
```
this seems sub-optimal. to really make Filesystem.File as functional and fast as IOStream, we need to copy the implement the in-Julia read/write buffers from the other LibuvStream subtypes.

samoconnor replied:
> this seems sub-optimal
I agree, I've [replaced it](https://github.com/samoconnor/julia/commit/b2eeae0aa58a91f7708a6a771465fab882e4aec8#diff-fe07d64cf361d7f56301455da5d58544R170) with `eof(f::File) = nb_available(f) == 0`.

This is not _optimal_ either but at least it is obviously a simple default implementation (i.e. open to further optimisation).

> to really make Filesystem.File as functional and fast as IOStream, we need to  ...

As I see it there are two (or three) steps:

1) Improve test coverage of `::IO` read/write to ensure that current implementations have consistent behaviour (even if some methods for some streams are not optimised). This is important so that later optimisations for particular stream types don't break behavioural compatibility.

2) Implement a pure-Julia buffering layer on top of `File` that makes it as fast/faster than `IOStream`. I think the buffering layer should be seperate from `File` and it should work on top of any low-level IO interface.

3) (maybe) Merge `IOStream` implantation with "`BufferedFile`" to remove dependancy on `support/ios.c` At that point `support/ios.c` would be used only by flisp.

tkelman replied:
> I've replaced it with `eof(f::File) = nb_available(f) == 0`.

Without deleting this version of the exact same method. One of them is redundant.

samoconnor replied:
whoops. fixed

samoconnor committed:
more tests for readall and readbytes

samoconnor commented:
Thanks to @mmaxs for https://github.com/JuliaLang/julia/pull/14630 which found a `LibuvStream` gap in the coverage of `test/read.jl`. This PR (#14699) now incorporates a revised version of @mmaxs fix from #14630.

samoconnor commented:
Windows now passing CI.
Linux i686 is passing CI.
Linux x86_64 is failing early in the build with [`make[1]: *** [julia-deps] Error 2`](https://travis-ci.org/JuliaLang/julia/jobs/103527838#L3455)

tkelman commented on test/read.jl:
```diff
@@ -0,0 +1,245 @@
+    end)
+end
+
+if !haskey(ENV, "TRAVIS")
```
If this can't run on travis, chances are it can't run on distro buildbots either. Not a good idea to special-case for CI like this.

samoconnor replied:
Most definitely agree!
Just attempting to discover which part of the code is causing deadlock...

tkelman replied:
ahk. did you track it down? I'm finding TCPSockets kind of challenging to add tests for myself. This branch helps get MbedTLS working again, though I'm not sure what its impacts are on the optimizations Jameson had been working towards.

samoconnor replied:
I hope that [this `yield()`](https://github.com/samoconnor/julia/commit/490cee37d8962d01dfabbcc719ff4062be3bb1cc#diff-435745fdfa02a0dc4f3b747ba977af11R63) will fix the deadlock. Waiting for Travis to start up to find out...

tkelman commented:
the commit https://github.com/samoconnor/julia/commit/29e5455792b2b74a8d2b38989c0494599f566012 that has nothing to do with File would be better suited for its own separate pr.

samoconnor committed:
update per #14660, add tests for filename-as-1st-arg read methods

tkelman commented:
remaining untested methods from #14608:

- `readlines` of a filename
- `readuntil` of a filename
- `eachline` of a filename
- `write(::IO, ::IO)` for IO types that are not subtypes of AbstractIOBuffer

samoconnor committed:
write(::IO, ::IO) tests

samoconnor committed:
Disable test of read*() from open(::Command).

https://github.com/JuliaLang/julia/issues/14747

clean up (hopefully) redundant celanup() calls in test/read.jl (left over from debugging hanging CI)

tkelman commented on test/read.jl:
```diff
@@ -0,0 +1,314 @@
+    @test read(io(), UInt8) == read(filename, UInt8)
+    @test read(io(), Int) == read(IOBuffer(text), Int)
+    @test read(io(), Int) == read(filename,Int)
+                                                                       cleanup()
```
the strange spacing is really distracting here

samoconnor replied:
I've just taken out most of the temporary `cleanup()` calls.
That was an attempt to figure out why CI was hanging (i.e. maybe because of too many tasks/file handles/sockets ...).

samoconnor commented:
> @tkelman: untested newly introduced methods: readavailable

You're talking about these right?
```julia
readavailable(io::File) = read(io)
readavailable(io::AbstractIOBuffer) = read(io)
readavailable(s::IOStream) = read!(s, Vector{UInt8}(nb_available(s)))
```
... called from here:
```julia
function write(to::IO, from::IO)
    while !eof(from)
        write(to, readavailable(from))
    end
end
```

...which is tested with 15 combinations of `to` and `from` stream types [here](https://github.com/samoconnor/julia/commit/4ed08cde7328b58770678271fbfb9e60f2f0dcd9#diff-435745fdfa02a0dc4f3b747ba977af11R291).

What, in your opinion, would be an adequate test???

tkelman commented:
That's fine. An indirect test is better than no test, but that's literally the only call site of `readavailable` other than one line in `examples/clustermanager/0mq/ZMQCM.jl`.

samoconnor commented:
Looks like all the CI is passing now. (with #14747 test disabled).

Passing tests cover something like:
    {filename, File, IOStream, IOBuffer, TCPSocket, PipeEndpoint}
    **X** {read(s, T), read!(s) read!(s, a), eof(s), readstring(s), read(s), read bytes!(s, a), readuntil(s,c), readline(s), readlines(s), eachline(s), countlines(s), readcsv(s), position(s), seek(s,n), skip(s), seekend(s)} 
    **X** {short CSV string,  + 4 x longer test strings approx SZ_UNBUFFERED_IO in length.}

samoconnor commented:
> ... that's literally the only call site of readavailable other than ...

I think I understand your concern. I think that `readavailable` should eventually be deprecated (or not exported).
See https://github.com/JuliaLang/julia/pull/14546#issuecomment-169856007 and https://github.com/JuliaLang/julia/pull/7478.

I also think that this generic implementation of `write(to::IO, from::IO)` should almost never be used once appropriate stream-type specialised methods are added. e.g. specialised methods should use `sendfile` in some cases or maybe `mmap` in others.

tkelman commented:
It is pretty widely used in packages according to github code search, which is most of my concern about it not being directly unit-tested here. So un-exporting it or deprecating it would break code outside of base but not the tests within base (still probably worth the cleanup to do at some point though). "untested" was imprecise and unfair wording, sorry.

samoconnor commented:
ready to merge?

vtjnash commented:
> It is pretty widely used in packages according to github code search...

that's annoying. i just read through the 12 pages of search results and found the following use cases for `readavailable`:
1) a few were correct reimplementations of sendfile (aka `write(to::IO, from::IO)`), half of the time `to` is another real IO object, the other half of the time, it's a pseudo-IO / display object (like a WebSocket or GUI console)
2) the rest were examples of broken code that might hang or print incomplete fragments if the output was non-triviala
